File: duna/Boj1654.py
import sys
k,n=map(int,input().split())  #이미 가지고 있는 k, 필요한 n
lan=[int(input()) for _ in range(k)]

def is_valid(cm):   #해당값 나눈 길이 계수 n 이상인지 판단
  cnt=0
  for i in range(k):
    if lan[i]>=cm:
      cnt+=lan[i]//cm
  
  return cnt>=n
   

left=1
#right=2147483647  #2^31-1, 최대값
right=max(lan)
answer=0

while left<=right:
  mid=(left+right)//2

  if is_valid(mid):
    answer=mid
    left=mid+1

  else:
    right=mid-1

print(answer)

File: changhyun/백준문제/Boj1654.java
package 백준문제;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Boj1654 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int K, N;
    static int[] wires;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        K = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());

        wires = new int[K];

        int maxLength = Integer.MIN_VALUE;
        for (int i = 0; i < K; i++) {
            wires[i] = Integer.parseInt(br.readLine());
            maxLength = Math.max(maxLength, wires[i]);
        }

        long left = 1;
        long right = maxLength;

        while (left <= right) {
            long mid = (left + right) / 2;

            int cnt = 0;
            for (int wire : wires) {
                cnt += wire / mid;
            }

            if (cnt < N) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(right);

    }
}
